### ISSUE TITLE
NoSuchMethodError: mpicbg.models.Boundable.estimateBounds

### ISSUE THREAD

**Issue by psteinb** [bigdataviewer/SPIM_Registration#10]
Hi to all,

I wanted to do some timing measurements again and couldn't get SPIM_Registration to perform a multi-view deconvolution. I am using 0d78e4c133340cbfbdd02c6083884fb383eb40e4 and build the package from the command line with ```mvn```.

```
Populating OME metadata
Opening '/scratch/steinbac/multiview_data/celegans/spim_TL1_Angle0.tif' [1388x1040x67 type=uint16 image=Image<FloatType>]
LOCI.openLOCI(): Cannot read metadata, setting calibration to 1
There was a problem with the class mpicbg.models.Boundable which can be found here:
/home/steinbac/software/fiji/edited/jars/mpicbg-0.6.1.jar
java.lang.NoSuchMethodError: mpicbg.models.Boundable.estimateBounds([D[D)V
        at mpicbg.imglib.algorithm.transformation.ImageTransform.<init>(ImageTransform.java:98)
        at mpicbg.spim.postprocessing.deconvolution.ExtractPSF.transformPSF(ExtractPSF.java:387)
        at mpicbg.spim.postprocessing.deconvolution.ExtractPSF.extract(ExtractPSF.java:279)
        at mpicbg.spim.fusion.PreDeconvolutionFusionSequential.fuseSPIMImages(PreDeconvolutionFusionSequential.java:555)
        at mpicbg.spim.fusion.FusionControl.fuse(FusionControl.java:80)
        at mpicbg.spim.fusion.FusionControl.fuse(FusionControl.java:23)
        at mpicbg.spim.Reconstruction.processIndividualViewStructure(Reconstruction.java:330)
        at mpicbg.spim.Reconstruction.<init>(Reconstruction.java:44)
        at fiji.plugin.Multi_View_Deconvolution.run(Multi_View_Deconvolution.java:83)
        at ij.IJ.runUserPlugIn(IJ.java:199)
        at ij.IJ.runPlugIn(IJ.java:163)
        at ij.Executer.runCommand(Executer.java:131)
        at ij.Executer.run(Executer.java:64)
        at java.lang.Thread.run(Thread.java:662)
```
I am using SPIM_Registration with plain .tif files from a/the celegans dataset that Stephan gave me. the plugin stops working after that. the fiji log console's last lines are:

```
Initialize combined weighteners for all views (Blending)
Computing transformed input views ... 
(Fri Mar 13 09:10:45 CET 2015): Opening '/scratch/steinbac/multiview_data/celegans/spim_TL1_Angle0.tif' [ArrayContainerFactory]
Extracting PSF for spim_TL1_Angle0.tif
PSF size: (17, 17, 29)

```

Anybody have an idea, what the problem is? I'm wondering, if SPIM_Registration as a plugin supported anymore as there is increasing uptake of SPIM Reconstruction.

Best,
P

**Comment by dscho** [comment]
@psteinb please do not leave out the installation step from your description (compare to http://stackoverflow.com/help/mcve and http://whathaveyoutried.com/, listed in the really, really helpful http://imagej.net/Bug_reporting_best_practices page that I highly recommend you should read if you haven't done so already).

The exception you reported indicates that there is a version skew. You will first have to find out in which of the build dependencies that method (i.e. that class) resides, then have to compare that dependency's version against the one you have in the setup where you are running the reconstruction.

See http://fiji.sc/Frequently_Asked_Questions#Why_do_I_get_a_NoSuchMethodError.2FNoSuchClassDefFoundError_when_running_a_rebuilt_plugin.3F for more information.

> I'm wondering, if SPIM_Registration as a plugin supported anymore

If you intend to rely on the SPIM Registration it might be a wise idea to think about ways to give back for what you ask for, so that it is a fair exchange between you and the developers who help you.

**Comment by psteinb** [comment]
thanks @dscho for jumping on this,

On 03/13/2015 11:04 AM, dscho wrote:
> @psteinb please do not leave out the installation step from your description (compare to http://stackoverflow.com/help/mcve and http://whathaveyoutried.com/, listed in the really, really helpful http://imagej.net/Bug_reporting_best_practices page that I highly recommend you should read if you haven't done so already).
>
apologies, I am a bit in a hurry to finish:
https://github.com/psteinb/libmultiviewnative
for a conference next week.
thanks for the links, I'll have a look!

> The exception you reported indicates that there is a version skew. You will first have to find out in which of the build dependencies that method (i.e. that class) resides, then have to compare that dependency's version against the one you have in the setup where you are running the reconstruction.
>
> See http://fiji.sc/Frequently_Asked_Questions#Why_do_I_get_a_NoSuchMethodError.2FNoSuchClassDefFoundError_when_running_a_rebuilt_plugin.3F for more information.
>

Mmmh, I just checked ... the same error occurs with a plain vanilla fiji 
install from the tarball downloaded this morning (virgin plugins 
directory). the shipped SPIM_Registration jar lists versions
SPIM_Registration-2.1.2.jar, so it might not only be my self-made plugin.

>> I'm wondering, if SPIM_Registration as a plugin supported anymore
>
> If you intend to rely on the SPIM Registration it might be a wise idea to think about ways to give back for what you ask for, so that it is a fair exchange between you and the developers who help you.

on giving back:
https://github.com/psteinb/libmultiviewnative
;)
which I am working on for quite some time now. As soon as it is ready, 
I'll submit PRs.

**Comment by dscho** [comment]
> Mmmh, I just checked ... the same error occurs with a plain vanilla fiji install from the tarball downloaded this morning (virgin plugins directory). the shipped SPIM_Registration jar lists versions SPIM_Registration-2.1.2.jar, so it might not only be my self-made plugin.

As I said, it is probably a version skew *in a dependency*. From the name, I would strongly suspect [mpicbg](https://github.com/axtimwalde/mpicbg) to be the culprit, as `SPIM_Registration` [does not even have a `Boundable` class](https://github.com/bigdataviewer/SPIM_Registration/find/master/) while `mpicbg` [does](https://github.com/axtimwalde/mpicbg/blob/master/mpicbg/src/main/java/mpicbg/models/Boundable.java).

The exception even mentions that there was a problem with mpicbg in version 0.6.1:

> ```
> There was a problem with the class mpicbg.models.Boundable which can be found here:
> /home/steinbac/software/fiji/edited/jars/mpicbg-0.6.1.jar
> ```

The 0.6.1 tag is unfortunately missing (in fact, `mpicbg` has no tags at all), but you can see easily that it was released with [this commit](https://github.com/axtimwalde/mpicbg/commit/d14548dec4d4564542c50a66b24429dd86735ea2). Looking at the `Boundable` interface from that revision, you see that [it has a different `estimateBounds` signature](https://github.com/axtimwalde/mpicbg/blob/d14548dec4d4564542c50a66b24429dd86735ea2/mpicbg/src/main/java/mpicbg/models/Boundable.java#L19).

So I was right all along.

**Comment by axtimwalde** [comment]
You are working with a SPIM_Registration that depends on mpicbg-1.0.0.
Use an older SPIM_Registration or pull in mpicbg-1.0.0 which may break
other things but should work for what you want to do.  Are you sure that
you got this SPIM_Registration with an official update/ download?

mpicbg-1.0.0 isn't in the updates or downloads yet because not all of
its dependents have been upgraded.  I am working on this whenever I find
time.  And it needs testing...

Best,
Stephan

**Comment by ctrueden** [comment]
Firstly, thanks to @dscho for updating the ImageJ FAQ about version skew diagnosis!

The issue tracking `mpicbg` upgrade progress in Fiji can be followed here: https://github.com/fiji/fiji/issues/113

@psteinb is correct that there is a version skew with the current version of Fiji. Unfortunately, [this commit of legacy-imglib1](https://github.com/fiji/legacy-imglib1/commit/cb773c612a92f28923c8cb3f01de3b167e6e0edc) force-updated the version of `mpicbg` to `1.0.0` while releasing it as `1.1.0-DEPRECATED`. Then [this commit of pom-fiji](https://github.com/fiji/pom-fiji/commit/ecb27feb10eafc4305c9ec1efded4a88efa43554) bumped the `legacy-imglib1` version to `1.1.0-DEPRECATED`. Then `legacy-imglib1` version `1.1.0-DEPRECATED` was subsequently uploaded to the Fiji update site, but without a corresponding upload of `mpicbg` version `1.0.0` (since as @axtimwalde says, there are other components which have not yet been upgraded to use it).

That release of `legacy-imglib1` _should_ have been `2.0.0-DEPRECATED` since it was backwards incompatible with the previous version. Because it was not labeled as a major version, @hinerm updated the version in `pom-fiji` thinking it was backwards compatible by SemVer.

To address this problem, I have [reverted the problematic commit of `legacy-imglib1` on master](), and will release `legacy-imglib1` version `1.1.1-DEPRECATED` fixing this fatal backwards incompatibility. I started a [topic branch of `legacy-imglib1` that contains the mpicbg-1.0.0 updates](https://github.com/fiji/legacy-imglib1/compare/mpicbg-1.0.0), which can be merged to master once we are ready to release version `2.0.0-DEPRECATED` in conjunction with the other `mpicbg`-dependent Fiji components.

**Comment by dscho** [comment]
I see. It is a very unhappy situation.

**Comment by ctrueden** [comment]
* Released [legacy-imglib1 1.1.1-DEPRECATED](https://github.com/fiji/legacy-imglib1/tree/legacy-imglib1-1.1.1-DEPRECATED)
* Uploaded it to the Fiji update site
* Rebased the [legacy-imglib1 ome-5.1 branch](https://github.com/fiji/legacy-imglib1/compare/ome-5.1) over the latest master
* Uploaded the OME 5.1 version of legacy-imglib1 to the Bio-Formats update site

So now this skew should be gone in vanilla Fiji, as well as in Fijis with the Bio-Formats update site enabled.

No change is necessary in `SPIM_Registration` itself because the version skew is not present in its actual dependencies on `master`: it depends on `legacy-imglib1` version `1.0.0-DEPRECATED` which depends on `mpicbg 0.6.1` so all is OK for the moment.

**Comment by axtimwalde** [comment]
@ctrueden Thank you very much for fixing this issue!  I am very sorry that the pending update of mpicbg has lead to this confusion.  However, I have to argue against the statement that legacy-imglib1 should have been 2.0.0-DEPRECATED.  The public legacy-imglib1-1.1.0-DEPRECATED is strictly backwards compatible plus enhancements which according to SemVer leads to an increment in the second position which I did.  I took care of this very carefully.  The problem underlying the current confusion is that legacy-imglib1-1.1.0-DEPRECATED was distributed together with an incorrect dependency which is the result of including it in pom-fiji as is although it overrides a dependency version.  This is the event that has to be prevented in the future.  I understand that overriding dependency versions of managed dependencies of pom-fiji makes things more complicated for maintaining pom-fiji, but it is necessary to not freeze development.  I could not have taken care of everything depending on mpicbg at once---I did a lot already, but I ran out of time.  I hope to find time soon to do the rest.  Thanks again for your help resolving this!

**Comment by axtimwalde** [comment]
More reading on the dependency versioning cascade discussion that concur with my interpretation of SemVer, but leave room because it is purely specified

http://semver.org/

http://stackoverflow.com/questions/23096170/should-major-semver-updates-cascade

https://github.com/mojombo/semver/issues/148

I can see that in Java the dominating conclusion (which is in line with what I did) may be questioned because there is no such thing as a private dependency.  So, I suggest that since we now ran into this issue, we should amend our guidelines with: "Dependencies following SemVer bubble their versions." or more precisely "The SemVer version of an API component used in Fiji increases at the highest or higher order of the SemVer version of any of its dependencies.".  That would be compatible with @ctrueden 's fix of imglib1-legacy and will enable us to reliably accuse violators in the future (this was meant to be funny, no offense).  What do you think?  Where should we post this?

**Comment by axtimwalde** [comment]
Thinking about it longer I am in doubt.  The downside of this proposed bubbling  scheme is that version numbers will run fast and *do not reflect changes in the API of the component* itself which in my understanding is the primary intention of SemVer.  Consumers of a library facing a version change from 1.0.0 to 2.0.0 will expect dramatic API change including incompatibilities and may feel reluctant to adapt although the cause for the version increase was a `private' dependency.  The proposed amendment serves only the Fiji distribution mechanism in its current form which may be addressed by other mechanisms better (i.e. tracing the dependency versions through the pom dependency-tree and identifying inconsistencies).  I'd be thankful for thoughts on this.  I think this is worth discussing.

**Comment by ctrueden** [comment]
@axtimwalde All your comments are right on—I have thought about the pros and cons of SemVer cascades as well and I am ambivalent. We should discuss on the imagej-devel mailing list. I'll start the thread when I'm back in the office on Monday, if some one else does not do so first.

**Comment by axtimwalde** [comment]
Thanks @ctrueden!  BTW, I just found that mpicbg-1.0.0 has a small yet fundamental bug that prevented it from doing its very core work, robust model fitting, i.e. 1.0.0 is a dysfunctional release including everything that depends on it.  I pushed a fix with version number 1.0.1, and we will have to update the dependencies later.  Too tired for that now :(.

**Comment by dscho** [comment]
> The public legacy-imglib1-1.1.0-DEPRECATED is strictly backwards compatible

Since it relies on a different `mpicbg` version (which is not only backwards-incompatible but also forward-incompatible to the previous version), that is incorrect. You would not be able to drop in *just* the `legacy-imglib1-1.1.0-DEPRECATED.jar` file *and keep `mpicbg-0.6.1.jar`*; that would already break it.

This is not a new problem: @hinerm, @ctrueden and myself dealt with these issues in the ImgLib2 context all the time.

From a theoretical perspective, the most appropriate solution would be to switch to OSGi. From a practical perspective, however, it should be obvious that OSGi would not be accepted by the community: just look at the opposition to Maven (which is comparatively dirt simple). While some have picked it up in the meantime, others have stated publicly that they will never use Maven (even if some of them acknowledge the advantages).
